bentoml/_internal/frameworks/tensorflow_v2.py

```diff
@@ -386,129 +183,93 @@ def __call__(self, inputs):
-        device_id (:code:`str`, `optional`, default to the first CPU):
-            Optional devices to put the given model on. Refers to `Logical Devices <https://www.tensorflow.org/api_docs/python/tf/config/list_logical_devices>`_ from TF documentation.
-
-    Returns:
-        :obj:`~bentoml._internal.runner.Runner`: Runner instances for :mod:`bentoml.tensorflow` model
-
-    Examples:
 
-    .. code-block:: python
+    partial_kwargs: t.Dict[str, t.Any] = bento_model.info.options.get(
+        "partial_kwargs", dict()
+    )
 
-        import bentoml
+    class TensorflowRunnable(Runnable):
+        SUPPORT_GPU = True
+        SUPPORT_MULTI_THREADING = True
 
-        # load a runner from a given flag
-        runner = bentoml.tensorflow.load_runner(tag)
+        def __init__(self):
+            super().__init__()
+            if len(tf.config.list_physical_devices("GPU")) > 0:
+                self.device_name = "/device:GPU:0"
+            else:
+                self.device_name = "/device:CPU:0"
+                num_threads = int(os.environ["BENTOML_NUM_THREAD"])
+                tf.config.threading.set_inter_op_parallelism_threads(num_threads)
+                tf.config.threading.set_intra_op_parallelism_threads(num_threads)
+
+            self.model = load_model(bento_model, device_name=self.device_name)
+            self.methods_cache: t.Dict[str, t.Callable[..., t.Any]] = {}
+
+    def _gen_run_method(runnable_self: TensorflowRunnable, method_name: str):
+        raw_method = getattr(runnable_self.model, method_name)
+        method_partial_kwargs = partial_kwargs.get(method_name)
+        if method_partial_kwargs:
+            raw_method = functools.partial(raw_method, **method_partial_kwargs)
+
+        def _run_method(
+            runnable_self: TensorflowRunnable, *args: "TFArgType", **kwargs: "TFArgType"
+        ) -> "ext.NpNDArray":
+            params = Params["TFArgType"](*args, **kwargs)
+
+            with tf.device(runnable_self.device_name):
+
+                def _mapping(item: "TFArgType") -> "tf_ext.TensorLike":
+                    if not LazyType["tf_ext.TensorLike"]("tf.Tensor").isinstance(item):
+                        return t.cast("tf_ext.TensorLike", tf.convert_to_tensor(item))
+                    else:
+                        return item
+
+                params = params.map(_mapping)
+                res = raw_method(*params.args, **params.kwargs)
+                return t.cast("ext.NpNDArray", res.numpy())
+
+        return _run_method
+
+    def add_run_method(method_name: str, options: ModelSignature):
+        def run_method(
+            runnable_self: TensorflowRunnable, *args: "TFArgType", **kwargs: "TFArgType"
+        ) -> "ext.NpNDArray":
+            _run_method = runnable_self.methods_cache.get(method_name)
+            if not _run_method:
+                _run_method = _gen_run_method(runnable_self, method_name)
+                runnable_self.methods_cache[method_name] = _run_method
```

Comment:
Shouldn't we do this at init time and not at runtime?

Reply:
I still feel putting this part in the first time a method is ran is more nature with little performance penalty. We can change it later if necessary.